### ISSUE TITLE
Documentation broken or source reference missing?

### ISSUE THREAD

**Issue by Helveg** [mpi4py/mpi4py#26]
None of the class links in the doc, such as `MPI.Comm` work (but intersphinx links such as `pickle` do work).

On top of that there also doesn't seem to be a way to view the reference docs or source code docs. Is the doc build broken?

**Comment by dalcinl** [comment]
The class links are not there in the docs, I never managed to implement it. In the past I would have sent you to the Epidoc-generated API reference [here](https://mpi4py.github.io/apiref/index.html), but now that Python 2 is definitely dead, I should really have all docs generated by Sphinx. I'm working sparingly on it, there are a few different options (e.g. autodoc and autoapi), and I still have to settle with one. Suggestions are most welcome.

mpi4py documentation is definitely not great, but to my defense, the MPI API (and thus mpi4py) is rather large to be written by a one-man team.

**Comment by Helveg** [comment]
I've tried looking into the source code before but I'm unfamiliar with anything that isn't pure Python. If it's as simple as adding docstrings and using Sphinx's [autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) then I can definitely contribute.

**Comment by dalcinl** [comment]
Well, what's primarily left is to use autodoc (or something similar) on the `mpi4py.MPI` module. This is an extension modules written in Cython. Docstrings are already there, but they are all one-line, no description of parameters, returns, etc. at all. It is about 500 fuctions+methods, improving on that requires a lot of time, and some knowledge of MPI. Besides improving docstrings, what mpi4py is needing is a way to automatically generate the API reference, I've tried sphinx-automodule and sphinx-autosummary, as well as [sphinx-automodapi](https://sphinx-automodapi.readthedocs.io/) and [autoapi](https://autoapi.readthedocs.io/), and I'm still not sure in which basket I should put the eggs.

**Comment by bdice** [comment]
This would be a great issue to work on! I would be in favor of an API page that lists all the functions/classes and their signatures. I was able to learn what I needed by the example snippets in the docs, combined with StackOverflow and some prior knowledge of MPI. However, I think I could have answered my own questions more quickly if there was a comprehensive API reference in the docs. I have done this for a few packages before, and I might be able to help with this at some point. I have a backlog of “open source contributions I want to make” so no promises on a timeline, but I would be happy to help review and/or assist if someone gets stuck.

**Comment by Helveg** [comment]
Same as @bdice, I can help out and probably even use it as an opportunity to learn about some more exotic functions. Maybe just pick 1 option @dalcinl to get us started and if it turns out it was the wrong choice we can deal with its shortcomings in the future. I'd say it's better to have some API docs now, than the perfect API docs in a future that might never come ;p

**Comment by bdice** [comment]
I have found this set of extensions to be a pretty good option:
https://github.com/glotzerlab/coxeter/blob/df2048a9451268389fe62d9ab9f8bbcd41afafa2/doc/source/conf.py#L34-L42

With `sphinx.ext.autodoc`, `sphinx.ext.autosummary`, and `autodocsumm`, you can transform a pretty minimal rST file like [this](https://github.com/glotzerlab/coxeter/blob/master/doc/source/package-shapes.rst) into an API reference like [this](https://coxeter.readthedocs.io/en/latest/package-shapes.html). It uses the module, class, and method docstrings to generate it all.

**Comment by bdice** [comment]
I spent a few minutes looking at the current docs infrastructure. I am not familiar with epydoc, and the structure of the documentation is rather different than the Sphinx projects I have worked with in the past. I'm not sure what to propose -- incremental changes within the current framework, or a total overhaul with a modern Sphinx configuration and just copy content from the rst files? Maintainers, what do you want to see for the package?

**Comment by dalcinl** [comment]
@bdice Forget about epydoc. It looks like we need a major overhaul, indeed. What we need is a way to document the `mpi4py.MPI` module, automatically if possible (although all docstrings are one-liners, nor really useful docs besides minimal descriptions ). 

`mpi4py.MPI` is an extension module, so some doc tools may struggle to handle it properly as introspection is obviously not available. For example, my quick trials failed to make Sphinx generate entries for module constants like `MPI.ANY_SOURCE` or built-in instances like `MPI.COMM_WROLD` or `MPI.DOUBLE`.

I spent a lot of time adding type annotations to all function and methods. Then I was able to generate typing stubs, look for mpi4py/MPI.pyi.  Ideally, auto-generated docs should expose type hints in function and method signatures. Maybe we should write some sort of script that traverses the `mpi4py.MPI` module and generates *.rst sources (sort of our own custom autodoc)? That's the approach I took to generate typing stubs (mpi4py/MPI.pyi), take a look at `conf/mpistubgen.py`.

I just warn you all: Current autodoc approaches do a damn good job at generating docs for pure Python code. But once you feed them Cython extension modules, you may end up disappointed with the outcome. This is not a complaint on these awesome tools, those feeling entitled to complain should rather spend their energy contributing patches.

**Comment by bdice** [comment]
@dalcinl Excellent, I agree with your assessment. I maintain some Cython packages with C++ extension modules (https://freud.readthedocs.io/en/stable/) and the documentation can work fine as long as the package can be built with the right flags. I will take a look at this and see if mpi4py would work using the same approaches I’ve used before.

**Comment by Helveg** [comment]
I haven't taken a look yet at `mpistubgen.py` yet, but couldn't we extract all signatures and create stub python files a la:

```python
COMM_WORLD = None
"""The docstring for a constant"""

def Barrier(...):
  """ Incomplete docstring"""
   pass
```

We complete the docs in those files and then have sphinx base its apidocs on these pure Python files?

**Comment by dalcinl** [comment]
Yes, that should work, and what's in `mpistubgen.py` is very close to what you would need to generate a dummy pure Python `MPI` module to feed Sphinx with. You should start from `mpistubgen.py`, remove some typing-related stuff, and add docstrings to the output instead of throwing them away. Please note that the first line of docstrings in `mpi4py.MPI` are signatures generated with a custom Cython audodoc transform (`conf/cyautodoc.py`), that code should eventually be contributed back to Cython.
